source/opt/basic_block.h

```diff
@@ -106,38 +103,36 @@ class BasicBlock {
                              bool run_on_debug_line_insts = false);
 
   // Runs the given function |f| on each label id of each successor block
-  void ForEachSuccessorLabel(
-      const std::function<void(const uint32_t)>& f);
+  void ForEachSuccessorLabel(const std::function<void(const uint32_t)>& f);
 
   // Runs the given function |f| on the merge and continue label, if any
-  void ForMergeAndContinueLabel(
-      const std::function<void(const uint32_t)>& f);
+  void ForMergeAndContinueLabel(const std::function<void(const uint32_t)>& f);
 
  private:
   // The enclosing function.
   Function* function_;
   // The label starting this basic block.
   std::unique_ptr<Instruction> label_;
   // Instructions inside this basic block, but not the OpLabel.
-  std::vector<std::unique_ptr<Instruction>> insts_;
+  InstructionList insts_;
```

Comment:
Yes, I have thought of this.  However, I want to delay that decision until we have a better understanding of the trade-offs.  The task we want to accomplish is "Given an Instruction, find the basic block it is contained in."  If an instruction contains a pointer to the list it is contained in, then merging basic block in say code straightening, will be linear time: we have to traverse every instruction to update the pointer to the containing list.  However, it could make finding the basic block really fast.

To find the basic block, we can also make it such that we have to walk the basic block until we get to the label.  So finding the containing basic block is linear time, but merging two block is constant, just change a few pointers.

We'll have to decide what is more important.

Reply:
I agree.  Let's leave the decision for when we really need it.  Perhaps we could add a helper that does the instruction->block mapping on-demand.  We can factor out the code in #810 for that.